### PR TITLE
Add native near transfers Aurora -> Ethereum

### PR DESCRIPTION
**Pull request by kiseln**
In order to support this we make an XCC call to `migrate_to_ethereum`  method on NEAR. Required wNear is automatically attached to the XCC call, it just needs to be approved to the smart contract.

### Lack of callback

I didn't add an XCC callback, partly because I don't expect `migrate_to_ethereum` to ever fail. It can fail if the amount passed is 0, in which case user will not loose any funds. It can also fail if smart contract address passed is incorrect but that should never happen from our side.

### Creating an account for XCC calls

The implicit account for XCC calls will automatically be created during the first transfer. The caller needs to have additional 2 wNear to pay for that. I think the easiest solution is to make this part of deploy process and execute the first transfer ourselves.

### Upgradeability

Is this something we want to add?

### TODO

Add config for mainnet
Tests?

### CODE REVIEW

File: aurora/contracts/NearBridge.sol [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+contract NearBridge {
+    using AuroraSdk for NEAR;
+    using AuroraSdk for PromiseCreateArgs;
+    using AuroraSdk for PromiseWithCallback;
+
+    NEAR private near;
+    string private nearBridgeAccountId;
+
+    uint64 constant MIGRATE_TO_ETHEREUM_GAS = 150_000_000_000_000;
```
Comment (karim-en) [inline aurora/contracts/NearBridge.sol:17]: Please decrease it to 10TGAS

File: aurora/contracts/NearBridge.sol [hunk 2]
```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.8.24;
+
+import {AuroraSdk, NEAR, PromiseCreateArgs, PromiseWithCallback, Utils} from "@auroraisnear/aurora-sdk/aurora-sdk/AuroraSdk.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {Strings} from "@openzeppelin/contracts/utils/Strings.sol";
+import "./IEvmErc20.sol";
+
+contract NearBridge {
```
Comment (karim-en) [inline aurora/contracts/NearBridge.sol:9]: Please make the contract upgradable

**Review by karim-en — CHANGES_REQUESTED** [review]

File: aurora/contracts/NearBridge.sol (hunk 1 above)
Comment (kiseln) [reply aurora/contracts/NearBridge.sol]: Done in bf6e973a11ec52b5a64a6df295e4df089123bc95

File: aurora/contracts/NearBridge.sol (hunk 2 above)
Comment (kiseln) [reply aurora/contracts/NearBridge.sol]: Done in bf6e973a11ec52b5a64a6df295e4df089123bc95

File: aurora/README.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+Smart contract for transferring NEAR from Aurora to Ethereum. This is done by making an XCC call to the bridge contract with required wNear amount attached.
+
+**Attention**: we shall ourselves make the first transfer after deploying the contract in order to provide 2 NEAR required to create an account for XCC calls.
+
+## Working with the project
+
+Install dependencies
+```
+npm i
```
Comment (karim-en) [inline aurora/README.md:11]: Pleas use yarn instead of npm

**Review by karim-en — CHANGES_REQUESTED** [review]

File: aurora/README.md (hunk 3 above)
Comment (kiseln) [reply aurora/README.md]: Done d81def62d5045727d70c47c07990e1eec4c87bcc
I prefer yarn myself actually

File: aurora/contracts/NearBridge.sol [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    string private eNearAccountId;
+
+    uint64 constant MIGRATE_TO_ETHEREUM_GAS = 10_000_000_000_000;
+
+    event InitBridgeToEthereum(address indexed sender, address indexed recipient, uint128 amount);
+    
+    error AmountIsZero();
+    error RecipientIsZeroAddress();
+
+    function initialize(address wnear, string memory _eNearAccountId) external initializer {
```
Comment (karim-en) [inline aurora/contracts/NearBridge.sol:24]: It isn't strictly needed, but it is recommended to add `_disableInitializers` as a best practice.
More details is [here](https://forum.openzeppelin.com/t/what-does-disableinitializers-function-mean/28730) 
```
    // @custom:oz-upgrades-unsafe-allow constructor
    constructor() {
        _disableInitializers();
    }
```

File: aurora/README.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# Overview
+
+Smart contract for transferring NEAR from Aurora to Ethereum. This is done by making an XCC call to the bridge contract with required wNear amount attached.
+
+**Attention**: we shall ourselves make the first transfer after deploying the contract in order to provide 2 NEAR required to create an account for XCC calls.
+
+## Working with the project
```
Comment (karim-en) [inline aurora/README.md:7]: Please also add the verify command to verify the contract on `https://explorer.aurora.dev`

**Review by karim-en — CHANGES_REQUESTED** [review]

File: aurora/contracts/NearBridge.sol [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        near = AuroraSdk.initNear(IERC20(wnear));
+        eNearAccountId = _eNearAccountId;
+
+        __Ownable_init(msg.sender);
+        __Pausable_init();
+    }
+
+    function bridgeToEthereum(address recipient, uint128 amount) external whenNotPaused {
+        if (amount == 0) {
+            revert AmountIsZero();
```
[the remark this answers is not in the document]
Comment (kiseln) [reply aurora/contracts/NearBridge.sol:34]: Typed errors cost slightly less gas and reduce contract size. They are also part of the ABI so can be easier handled by the front-end. I think it's generally a recommended approach nowadays

File: aurora/ignition/aurora_tesnet.params.json [hunk 7]
```diff
@@ -0,0 +1,8 @@
+{
+    "Bridge": {
```
[the remark this answers is not in the document]
Comment (kiseln) [reply aurora/ignition/aurora_tesnet.params.json:2]: Yes it seemed a little weird for me as well but that's the default approach from the docs

**Review by olga24912 — CHANGES_REQUESTED** [review]

File: aurora/hardhat.config.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
+require("@nomicfoundation/hardhat-ignition-ethers");
+require("dotenv").config();
+
+const AURORA_PRIVATE_KEY = process.env.AURORA_PRIVATE_KEY;
+
+module.exports = {
+  solidity: "0.8.24",
+  networks: {
+    testnet_aurora: {
```
[the remark this answers is not in the document]
Comment (kiseln) [reply aurora/hardhat.config.js:11]: Done 4558f456b0257691151d760ceb81dec76db703d0

File: aurora/ignition/aurora_tesnet.params.json [hunk 9]
```diff
@@ -0,0 +1,8 @@
+{
```
[the remark this answers is not in the document]
Comment (kiseln) [reply aurora/ignition/aurora_tesnet.params.json]: Done 4558f456b0257691151d760ceb81dec76db703d0

File: aurora/README.md [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+Install dependencies
+```
+yarn
+```
+
+Build
+```
+npx hardhat compile
```
[the remark this answers is not in the document]
Comment (kiseln) [reply aurora/README.md]: Done 4558f456b0257691151d760ceb81dec76db703d0

File: aurora/README.md [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+```
+
+Deploy
+```
+npx hardhat ignition deploy ignition/modules/Bridge.js --network {} --parameters {insert_params_file}
```
[the remark this answers is not in the document]
Comment (kiseln) [reply aurora/README.md]: Done 4558f456b0257691151d760ceb81dec76db703d0

File: aurora/README.md (hunk 5 above)
Comment (kiseln) [reply aurora/README.md:7]: Added 4558f456b0257691151d760ceb81dec76db703d0

File: aurora/contracts/NearBridge.sol [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        if (recipient == address(0)) {
+            revert RecipientIsZeroAddress();
+        }
+
```
[the remark this answers is not in the document]
Comment (kiseln) [reply aurora/contracts/NearBridge.sol:45]: This is what I originally had but I've noticed then it transfers the tokens twice. Check this https://explorer.testnet.aurora.dev/tx/0xeafabbde871a2b6a5249ddf99b6fb7dddf7653831c990edb4e0bd2954951ac20

It is indeed different from what you would expect looking at examples. Shall we explore this further?

[the remark this answers is not in the document]
Comment (kiseln) [reply aurora/contracts/NearBridge.sol:45]: cc @karim-en

**Review by olga24912 — APPROVED** [review]

File: aurora/scripts/aurora_mainnet.params.json [hunk 13]
```diff
@@ -0,0 +1,8 @@
+{
+    "Bridge": {
```
Comment (karim-en) [inline aurora/scripts/aurora_mainnet.params.json:2]: @kiseln Why the format of mainnet and testnet configs are different?

**Review by karim-en — CHANGES_REQUESTED** [review]

File: aurora/scripts/aurora_mainnet.params.json (hunk 13 above)
Comment (kiseln) [reply aurora/scripts/aurora_mainnet.params.json]: My bad 1a179a6429df0e4afcdcfba85cc1f6acb8559d5f

File: aurora/contracts/NearBridge.sol (hunk 12 above)
[the remark this answers is not in the document]
Comment (paouvrard) [reply aurora/contracts/NearBridge.sol:45]: Could you please explain how this is working ? AFAIU the flow should be something similar to fast bridge:
- Pull the wNEAR token from the user wallet:
`wNEAR.transferFrom(msg.sender, address(this), amount);`
- Burn the wNEAR:
`wNEAR.withdrawToNear(bytes(getImplicitNearAccountIdForSelf(), ":unwrap"), amount);`
- Migrate the unwrapped NEAR to Ethereum:
`callMigrateToEthereum.transact();`

[the remark this answers is not in the document]
Comment (karim-en) [reply aurora/contracts/NearBridge.sol:45]: @paouvrard the XCC under the hood implements the first 2 action for payable calls, so the tokens are transferd to the implicit account id and then unwrapped.
The 3d action is explicitly called in the contract as you can see.

**Review by karim-en — APPROVED** [review]

File: aurora/hardhat.config.js [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+      accounts: [`0x${AURORA_PRIVATE_KEY}`],
+      chainId: 1313161555,
+    },
+    mainnet_aurora: {
+      url: "https://mainnet.aurora.dev",
+      accounts: [`0x${AURORA_PRIVATE_KEY}`],
+      chainId: 1313161554,
+    }
+  },
+  etherscan: {
```
Comment (karim-en) [inline aurora/hardhat.config.js:22]: The verification configuration  is missing for the Aurora network

Comment (karim-en) [reply aurora/hardhat.config.js:22]: 80bc8b7d2bd7570bfe2263a9c4cfb5c8cc115abb

**Pull request merged**